libkernelflinger/security.c

```diff
@@ -131,20 +132,58 @@ EFI_STATUS update_rot_data(IN VOID *bootimage, IN UINT8 boot_state,
                         return EFI_UNSUPPORTED;
         }
         rot_data.verifiedBootState = boot_state;
+
         temp_version.value = boot_image_header->os_version;
+        if (boot_image_header->header_version == BOOT_HEADER_V3) {
+                struct boot_img_hdr_v3 *boot_hdr = (struct boot_img_hdr_v3 *)bootimage;
+                temp_version.value = boot_hdr->os_version;
+        }
         rot_data.osVersion = (temp_version.split.version_A * 100 + temp_version.split.version_B) * 100 + temp_version.split.version_C;
-        rot_data.patchMonthYear = (temp_version.split.year + 2000) * 100 + temp_version.split.month;
+        /* VTS require the patchlevel's format should be YYYYMMDD, but the patch level's format in boot header */
+        /* is YYYYMM. We set the DD value to a fixed value of 1. */
+        rot_data.patchMonthYearDay = ((temp_version.split.year + 2000) * 100 + temp_version.split.month) * 100 + 1;
         rot_data.keySize = SHA256_DIGEST_LENGTH;
 
         if (vb_data) {
-                ret = rot_pub_key_sha256(vb_data, &temp_hash);
+                ret = rot_pub_key_sha256_and_avb_digest_type(vb_data, &temp_hash, &algorithm_type);
                 if (EFI_ERROR(ret)) {
                         efi_perror(ret, L"Failed to compute key hash");
                         return ret;
                 }
-                CopyMem(rot_data.keyHash256, temp_hash, rot_data.keySize);
+                if (state == LOCKED) {
+                        CopyMem(rot_data.keyHash256, temp_hash, rot_data.keySize);
+                } else {
+                        memset_s(rot_data.keyHash256, SHA256_DIGEST_LENGTH, 0, SHA256_DIGEST_LENGTH);
+                        barrier();
+                }
+
+                switch (algorithm_type) {
+                        /* Explicit fallthrough. */
+                        case AVB_ALGORITHM_TYPE_NONE:
+                        case AVB_ALGORITHM_TYPE_SHA256_RSA2048:
+                        case AVB_ALGORITHM_TYPE_SHA256_RSA4096:
+                        case AVB_ALGORITHM_TYPE_SHA256_RSA8192:
+                                avb_slot_verify_data_calculate_vbmeta_digest(
+                                                vb_data, AVB_DIGEST_TYPE_SHA256, rot_data.vbmetaDigest);
+                                rot_data.digestSize= AVB_SHA256_DIGEST_SIZE;
+                                break;
+                                /* Explicit fallthrough. */
+                        case AVB_ALGORITHM_TYPE_SHA512_RSA2048:
```

Comment:
algorithm_type is got from rot_pub_key_sha256_and_avb_digest_type(vb_data, &temp_hash, &algorithm_type); From the funciton name , the type should always sha256. So code can't reach SHA512, right?

Reply:
pub_key_sha256 and avb_digest_type are two different things. pub_key_sha256 means the public key will be hashed by using sha256, the result hash need to send to keymint. avb_digest_type are the digest type of the vbmeta data. Maybe the function name is a little confusing.